### PR TITLE
Add the ability to tell Shasta ATOM about exclusive

### PR DESCRIPTION
**Pull request by lisa-altair**
#### Describe Bug or Feature

Cray Shasta's ATOM interface has added an "exclusive" field as part of the ATOM REST interface for POSTing when a job starts.  The "exclusive" field will be set to true if either the job's -lplace= contains some form of "excl" or if the node the job will run on has a sharing value containing "excl".  To understand the outcome of the job/node combinations, please refer to the design.

#### Describe Your Change

I added logic to the PBS_cray_atom hook to correctly determine the value to pass to ATOM.

#### Link to Design Doc

https://pbspro.atlassian.net/wiki/spaces/PD/pages/1086029883/PBS+Design+Changes+for+Shasta

### CODE REVIEW

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 1]
```diff
@@ -176,6 +177,60 @@ class HookHelper(object):
...
+
+        if (sharing == pbs.ND_IGNORE_EXCL):
+            return False
+
+        excls = re.findall(r'excl', str(place))
+        totalexcls = len(excls)
+        if totalexcls > 0:
+            if "group" in str(place):
+                # A group is in the place statement, is it excl 
+                output = re.search(r'(?<=group=)[^:$]*', str(place))
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:207]: I'm not sure what this regex does, what does the input/output look like?

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 2]
```diff
@@ -176,6 +177,60 @@ class HookHelper(object):
         log_with_caller(pbs.EVENT_DEBUG4, 'path is %s' % path)
         return path
 
+    def is_it_exclusive(job):
+        """
+        check to see if the job requested exclusive, or if the
+        nodes are marked exclusive.  This needs to be passed
+        to ATOM.
+        """
+        place = job.Resource_List["place"]
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:186]: you use str(place) everywhere you use place, so you might as well just str it here.

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 3]
```diff
@@ -176,6 +177,60 @@ class HookHelper(object):
         log_with_caller(pbs.EVENT_DEBUG4, 'path is %s' % path)
         return path
 
+    def is_it_exclusive(job):
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:180]: This needs to have a `@staticmethod` decorator

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 4]
```diff
@@ -176,6 +177,60 @@ class HookHelper(object):
...
+        # Uses the same logic as the scheduler (is_excl())
+        if ((sharing == pbs.ND_FORCE_EXCL) or (sharing == pbs.ND_FORCE_EXCLHOST)):
+            return True
+
+        if (sharing == pbs.ND_IGNORE_EXCL):
+            return False
+
+        excls = re.findall(r'excl', str(place))
+        totalexcls = len(excls)
+        if totalexcls > 0:
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:204]: can you flip this if, you can save a level of indentation

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 5]
```diff
@@ -176,6 +177,60 @@ class HookHelper(object):
...
+            return False
+
+        if ((sharing == pbs.ND_DEFAULT_EXCL) or
+            (sharing == pbs.ND_DEFAULT_EXCLHOST)):
+            return True
+
+        if (sharing == pbs.ND_DEFAULT_SHARED):
+            return False
+
+        pbs.logmsg(pbs.LOG_DEBUG, "It was none of the above cases,"
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:231]: I would change this to log_with_caller, and change the message, someone looking at the logs won't know what the above cases are.
You could add log messages before all the other returns to give info to the admin.

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY (hunk 1 above)
Comment (lisa-altair) [reply src/modules/python/pbs_hooks/PBS_cray_atom.PY:207]: The input would look like what can be in a place statement.  For example a place statement could be: -lplace=excl:group=myexcl

But overall this section can be made cleaner.  Let me try to simplify it.

**Comment by lisa-altair** [comment]
@vstumpf I made the changes you requested and took one of your suggested to simplify the logic.   I still need to try out a few scenarios, but thought you might look at the changes in the mean time.  Just in case you have more comments.  :)

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 6]
```diff
@@ -176,6 +177,43 @@ class HookHelper(object):
...
+            return False
+
+        if ((sharing == pbs.ND_DEFAULT_EXCL) or
+            (sharing == pbs.ND_DEFAULT_EXCLHOST)):
+            return True
+
+        if (sharing == pbs.ND_DEFAULT_SHARED):
+            return False
+
+        # It was none of the above cases return not exclusive
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:215]: You don't really need this, the code makes sense

File: src/modules/python/pbs_hooks/PBS_cray_atom.PY [hunk 7]
```diff
@@ -176,6 +177,43 @@ class HookHelper(object):
...
+        log_with_caller(pbs.EVENT_DEBUG4, "place is %s" % place)
+
+        # See if the node sharing value has exclusive
+        vn = pbs.server().vnode(pbs.get_local_nodename())
+        sharing = vn.sharing
+        log_with_caller(pbs.EVENT_DEBUG4, "The sharing value is %s type %s" %
+                        (str(sharing), str(type(sharing))))
+
+        # Uses the same logic as the scheduler (is_excl())
+        if ((sharing == pbs.ND_FORCE_EXCL) or (sharing == pbs.ND_FORCE_EXCLHOST)):
```
Comment (vstumpf) [inline src/modules/python/pbs_hooks/PBS_cray_atom.PY:197]: you don't need parens around your if conditions, i would remove them.

**Comment by lisa-altair** [comment]
@vstumpf and @bayucan I have made some changes.  Please have a look.

**Pull request merged**